### PR TITLE
Problem (CRO-592) Problem: lite client code not tested.

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- A separate crate client-mock to generate legitimate tendermint data types, to test lite client verification.
- Other misc changes.
  - Change DateTime to Time, Timespec from i64 to u64.
  - Remove `Default` trait implementation of `HDSeed`.
  - Some hash field on tendermint data type should be `Option`.

### CODE REVIEW

**Comment by leejw51crypto** [comment]
lgtm.
squash to one commit

File: client-core/src/hd_seed.rs [hunk 1]
```diff
@@ -9,7 +9,7 @@ use crate::types::AddressType;
 use crate::Mnemonic;
 
 /// Hierarchical Deterministic seed
-#[derive(Debug, Default, PartialEq, Decode, Encode)]
+#[derive(Debug, Clone, PartialEq, Decode, Encode)]
```
Comment (yihuang) [inline client-core/src/hd_seed.rs]: The commit message is typo and mis-leading, I think the `Default` trait on `HDSeed` is insecure, the `Clone` instance is needed by `client-mock`.

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
seems the integration test failed: https://travis-ci.org/crypto-com/chain/jobs/617052366#L614

**Comment by yihuang** [comment]
> seems the integration test failed: https://travis-ci.org/crypto-com/chain/jobs/617052366#L614

Fixed, need to add new crate into `.dockerignore`.

**Comment by tomtau** [comment]
bors try

**Comment by tomtau** [comment]
bors retry

File: chain-abci/src/app/app_init.rs [hunk 2]
```diff
@@ -443,119 +467,117 @@ impl<T: EnclaveProxy> ChainNodeApp<T> {
...
+        let mut validator_liveness = BTreeMap::new();
+        let mut validator_by_voting_power = BTreeMap::new();
+        let mut tendermint_validator_addresses = BTreeMap::new();
+
+        for (address, node) in nodes.iter() {
+            let mut validator = ValidatorUpdate::default();
+            let power = get_voting_power(&conf.distribution, address);
+            self.validator_voting_power.insert(*address, power);
+            validator_by_voting_power.insert((power, *address), node.clone());
+            validator.set_power(power.into());
```
Comment (tomtau) [inline chain-abci/src/app/app_init.rs]: why is it creating validator updates here when at the end, it does `ResponseInitChain::new()` which won't contain this?

File: chain-tx-validation/src/lib.rs [hunk 3]
```diff
@@ -359,7 +359,7 @@ pub fn verify_unbonding(
     account.unbond(
         maintx.value,
         extra_info.min_fee_computed.to_coin(),
-        extra_info.previous_block_time + i64::from(extra_info.unbonding_period),
+        extra_info.previous_block_time + u64::from(extra_info.unbonding_period),
```
Comment (tomtau) [inline chain-tx-validation/src/lib.rs]: perhaps use the type alias:

```suggestion
        extra_info.previous_block_time + Timespec::from(extra_info.unbonding_period),
```

File: chain-abci/src/app/app_init.rs (hunk 2 above)
Comment (yihuang) [reply chain-abci/src/app/app_init.rs]: It's only for validating `req.validators` is correct.

File: chain-tx-validation/src/lib.rs (hunk 3 above)
Comment (yihuang) [reply chain-tx-validation/src/lib.rs]: Done

**Comment by tomtau** [comment]
@yihuang -- sorry, the suggestion I wrote didn't have an import: 
```

362 \|         extra_info.previous_block_time + Timespec::from(extra_info.unbonding_period),
--
740 | \|                                          ^^^^^^^^ use of undeclared type or module `Timespec`

```

**Comment by yihuang** [comment]
> @yihuang -- sorry, the suggestion I wrote didn't have an import:
> 
> ```
> 
> 362 \|         extra_info.previous_block_time + Timespec::from(extra_info.unbonding_period),
> --
> 740 | \|                                          ^^^^^^^^ use of undeclared type or module `Timespec`
> ```

I already fixed that.

File: chain-abci/src/app/app_init.rs [hunk 4]
```diff
@@ -244,6 +247,27 @@ fn store_valid_genesis_state(genesis_state: &ChainNodeState, inittx: &mut DBTran
     inittx.put(COL_EXTRA, b"init_chain_state", &encoded);
 }
 
+fn compute_accounts_root(account_storage: &mut AccountStorage, accounts: &[StakedState]) -> H256 {
+    let mut keys: Vec<_> = accounts.iter().map(StakedState::key).collect();
```
Comment (leejw51crypto) [inline chain-abci/src/app/app_init.rs]: let mut keys = accounts.iter().map(StakedState::key).collect();

Comment (yihuang) [reply chain-abci/src/app/app_init.rs]: The `Vec<_>` is necessary here for `collect` to know the correct type.

**Pull request merged**